### ISSUE TITLE
Review 1: 7.11.23

### ISSUE THREAD

**Issue by eddiebergman** [Sohambasu07/HPO_Suite#1]
Sorry for the pretty large review in the end. Some are minor comments/improvements, some are a bit more critical. Seems to be going well, just want to try to eliminate some bugs earlier on or at least make you aware of them so when the system grows, they aren't harder to find.

---
### Minor adjustment to MFPBench wrapper

https://github.com/Sohambasu07/HPO_Suite/blob/d35e44588520b26d460298b829fc6bdea6293cef/benchmarks/mfp_bench.py#L13-L22

mfpbench has a more explicit way to get config and result keys :)

```python
    table = mfpbench.get("lcbench_tabular", task_id=task_id, datadir=data_dir)
    table_for_task = table.table
    return TabularBenchmark(
        name=f"lcbench_tabular_{task_id}",
        table=table_for_task,
        id_key="id",  # Key in the table to uniquely identify tasks
        config_keys=table.config_keys,  # Keys in the table that correspond to configs
        result_keys=table.result_keys,  # Key in the table that corresponds to the result
        fidelity_key=table.fidelity_key,  # Key in the table that corresponds to the fidelity (e.g. "epoch")
    )
```

---
### Type of `fidelity`

https://github.com/Sohambasu07/HPO_Suite/blob/d35e44588520b26d460298b829fc6bdea6293cef/hpo_glue/glu.py#L84

Just a note on the typing of `list[int]` here. In some benchmarks it might actually be a `float`, i.e. percentage of data. You could change this to `list[int] | list[float]`

---
### Config ID's

https://github.com/Sohambasu07/HPO_Suite/blob/d35e44588520b26d460298b829fc6bdea6293cef/hpo_glue/glu.py#L118-L125

Right idea for getting configs from the table but unfortunately it's a bit more difficult. Say for example you have an MLP with some hyperparameters `{"learning_rate": 0.03, "n_nodes": 100}` and you've evaluated it at fidelity `[1,2,3]`. There should be only one `config_id` here and not three.
For every config, we will have a result per fidelity.

A suggestion on how to fix this:

```python
def _get_all_configs(self) -> None:
    """Get all possible configs for the benchmark"""
    non_duplicated_rows = df[self.config_keys].drop_duplicates()
    
    # Sorting is important to make sure it's always consistent
    sorted_configs = df.sort_values(by=self.config_keys)
    
    configs_as_dicts = sorted_configs.to_dict(orient="records")
    configs = [Config(i, config) for config in configs_as_dicts]
    
    self.configs = all_configs
```

Or if you like one-liners

```python
def _get_all_configs(self) -> None:
    """Get all possible configs for the benchmark"""
    self.configs = [
    	Config(i, config)
    	for i, config in enumerate(
	        df[self.config_keys]
	        .drop_duplicates()
	        .sort_values(by=self.config_keys)
	        .to_dict(orient="records")
	    )
    ]
```

---
### Use of methods for construction in `__init__`

https://github.com/Sohambasu07/HPO_Suite/blob/d35e44588520b26d460298b829fc6bdea6293cef/hpo_glue/glu.py#L104-L105

It's not clear that these methods actually assign something to `self` and for someone looking up the
class to see what attributes are available, they may not know that `self.configs` is a thing or `self.fidelities` is a thing.

My recommendation to help this would be

```python
class TabularBenchmark:

    def __init__(...):
        ...
        
        self.configs = self._get_all_configs(table, keys=config_keys)
        self.fidelities = self._get_all_fidelities(table, key=fidelity_key)
        
    # Note here that the function doesn't need a lot of information so it's a bit
    # clearer to just pass in what it needs and not depend on self
    @classmethod
    def _get_all_configs(cls, table: pd.DataFrame, keys: list[str]) -> list[Config]:
        return ...
        
    @classmethod
    def _get_all_fidelities(cls, table: pd.DataFrame, key: str) -> list[int]:
        return ...
```

---
### Docstrings
https://github.com/Sohambasu07/HPO_Suite/blob/d35e44588520b26d460298b829fc6bdea6293cef/hpo_glue/glu.py#L39-L46

The plan will probably be to use `mkdocs` (what is used for mfpbench and other libs). To make the class render nicely in documentatin, all we have to do is change those comments to a docstring under the attribute, i.e.

```python
class Result:
    """The result of a query from a benchmark."""
    
    query: Query
    """The query that generated this result."""
    
    result: dict[str, Any] 
    """Everything returned by the benchmark for a given query."""
    # TODO: We will handle singling out what thing is the objective as a ProblemStatement

    def __init__(self, query: Query, result: dict[str, Any]):
        self.query = query
        self.result = result
```

---
### Optimizer input to `run()`

https://github.com/Sohambasu07/HPO_Suite/blob/d35e44588520b26d460298b829fc6bdea6293cef/hpo_glue/glu.py#L152-L164

The `__init__` signature takes in a `type[Optimizer]` which in some sense is good and myself and Neeratyoy had talked about this problem before. The issue however is you call it later as 
`optimizer(config_space=benchmark.configs, fidelity_space=benchmark.fidelities)`. This will work for your random search but will not work for SMAC as the arguments to construct a SMAC optimizer are different.

There are two potential remedies to this:

* Explicitly define what `__init__` needs to take in the `class Optimizer` (more rigid, less complicated)
* Change the definition of the `__init__` as below (less rigid, more complicated)
```python

# from
def __init__(self, optimizer: type[Optimizer], ...): ...

# to
from typing import Callable

def __init__(self, optimizer: Callable[[list[Config], list[int]], Optimizer], ...): ...
```

The benefit of the second approach is that you can pass in your `RandomSearch` class as before, i.e. I can call `RandomSearch` which will call it's `__init__` which is a `Callable[[list[Config], list[int]], RandomSearch]`, therefor satisfying the thing above.

With the second approach, I could also pass in some function which will generate the right optimizer for the given benchmark, not having to create a new `Optimizer` class each time.

For now, just go with the first approach but I imagine the second approach will be something we have to migrate to later. You'll see this problem more when implementing `SMAC`.

---
### `is_tabular`

https://github.com/Sohambasu07/HPO_Suite/blob/d35e44588520b26d460298b829fc6bdea6293cef/hpo_glue/glu.py#L158

We don't have to pass this in, we can figure this out in `GLUE` with
```python
is_tabular = isinstance(benchmark, TabularBenchmark) 
```

Secondly, here:

https://github.com/Sohambasu07/HPO_Suite/blob/d35e44588520b26d460298b829fc6bdea6293cef/hpo_glue/glu.py#L175

I'm not sure we need to pass `is_tabular` into the `ask`. This variable won't change during the `run()` and so it's likely that this can just be picked up by the `Optimizer` when it's constructed with much the same technique

https://github.com/Sohambasu07/HPO_Suite/blob/d35e44588520b26d460298b829fc6bdea6293cef/optimizers/random_search.py#L31-L40

```python
    def ask(self, config_id: Optional[str] = None) -> Query:
        """Ask the optimizer for a new config to evaluate."""
        
        # We are dealing with
        if isinstance(self.configs, list):
            config = random.choice(self.config_space)
            fidelity = random.choice(self.fidelity_space)
            return Query(config, fidelity)
            
        # We are dealing with a surrogate
        else:
            config = self.config_space.sample_configuration(1, seed=self.seed)
            
            # I'm not sure how to deal with the `id` here...
            # There's no real order as the order in which you get configs every seed
            # will differ.
            # Perhaps we can also have `Configs` without an id? No idea...
            return Query(Config(config_id, config))

```

---
### `Optional`

You can also do this :)

```python
# before
def f(x: Optional[int] = None) -> int: ...
    
# after
def f(x: int | None = None) -> int: ...
```

This helps a lot when you have nested types as it removes one nesting level, i.e. `Optional[list[dict[str, Any]]]` vs `list[dict[str, Any]] | None`. 

---
### Seeding with `random`
There are some slight issues with seeding

https://github.com/Sohambasu07/HPO_Suite/blob/d35e44588520b26d460298b829fc6bdea6293cef/optimizers/random_search.py#L26-L29

This line sets the **global** rng, and not just the rng for this model. This can be brittle as if somewhere else entirely, someone does `random.choice`, this will effect your own use of `random.choice`.

To remedy this:
``` 
def __init__(self, ..., seed: int | None = None):
        
    ...
    self.rng = random.Random(seed)
    
```

See: https://docs.python.org/3/library/random.html#random.Random

---
### Seeding with `ConfigSpace`

https://github.com/Sohambasu07/HPO_Suite/blob/d35e44588520b26d460298b829fc6bdea6293cef/optimizers/random_search.py#L21-L24

The issue here is that if you always pass the same seed, you will always get the same config back out. This would only show up once you do the surrogate benchmarks, or at least one with a continuous space where you sample from the `ConfigurationSpace` object directly, and not the list of configs.

A possible remedy:

```python
sample_seed = self.rng.randint(0, 2**31-1)  # Big number so we can sample 2**31-1 possible configs
config = self.config_space.sample_configuration(num_configs, seed=sample_seed)
```

Not really a major performance issue but you could extract `2**32-1` as a file-level constant so it's not always recalculated at each sample.

There is also [`sys.maxsize`](https://docs.python.org/3/library/sys.html#sys.maxsize) which defines this constant but it varies between 32-bit and 64-bit systems so it's better we just have our own constant.

Also another thing to be aware of is you can pass a seed to `ConfigurationSpace.set_seed()` to essentially seed the space. If this is done, you'll want to do `copy.deepcopy` when passing the space from GLUE to the optimizer so that if running one optimizer after another, they won't interfere.

### Nit-pick
https://github.com/Sohambasu07/HPO_Suite/blob/d35e44588520b26d460298b829fc6bdea6293cef/hpo_glue/glu.py#L194-L198

```python
cols = (
    ["Trial", "Optimizer", "Benchmark", "Config id", "Fidelity"]
    + benchmark.config_keys
    + benchmark.result_keys
)
```

In the future, we probably just want some object, i.e. `History` which has a method

```python
class History:
    results: list[Result]
    
    def __init__(self):
        self.results = []
    
    def add(self, result: Result) -> None:
        self.results.append(result)
        
    def df(self) -> pd.DataFrame:
        # Convert to a dataframe here. You can extract the columns from
        # the `Query`, `Config` and `Result`
```

To add in the benchmark name and things, that can be done by the glue layer in the `run()`

```
history = History()
while ...:
   ...
   history.add(result)
   
df = history.df()
df["benchmark"] = benchmark.name
df["optimizer"] = optimizer.name
```

---
### Saving results

https://github.com/Sohambasu07/HPO_Suite/blob/d35e44588520b26d460298b829fc6bdea6293cef/hpo_glue/glu.py#L202-L209

I imagine this logic should live outside the `GLUE` part, i.e. in the experiments section. GLUE's responsiblities is just to run and collect information and return it. Not a problem for now but just to have it written down somewhere.

---
### `TabularBenchmark.query()`

https://github.com/Sohambasu07/HPO_Suite/blob/d35e44588520b26d460298b829fc6bdea6293cef/hpo_glue/glu.py#L107-L116

Not all tabular benchmarks are going to have their tables neatly formatted unfortunatly :( You'll probably have to use the parameters and table in `__init__` to make sure all tables follow the same format such that the `query` function works for all tables. The `mfpbench.TabularBenchmark` does this so I believe you can copy some code from there.
